### PR TITLE
[BUGFIX beta] Make substates great again!

### PR DESCRIPTION
**Pull request by trentmwillis**
This PR is a pretty substantial overhaul of the internals for how we handle loading/error substates and routes. This makes it much easier (IMO) to follow what is going on internally.

It also fixes issues with Engines, such as https://github.com/dgeb/ember-engines/issues/193. To verify, it introduces a slew of new tests and removes some old ones that weren't testing the behavior we actually wanted.

### CODE REVIEW

File: packages/ember-glimmer/tests/integration/application/engine-test.js [hunk 1]
```diff
@@ -124,6 +139,220 @@ moduleFor('Application test: engine rendering', class extends ApplicationTest {
...
   }
 
+  ['@test error substate route works for the application route of an Engine'](assert) {
+    this.setupAppAndRoutableEngine();
+    this.application.__registry__.resolver.moduleBasedResolver = true;
+    this.additionalEngineRegistrations(function() {
+      this.register('template:application_error', compile('Error!'));
+      this.register('route:post', Route.extend({
+        model() {
+          return RSVP.reject();
```
Comment (rwjblue) [inline packages/ember-glimmer/tests/integration/application/engine-test.js]: Can you reject with an Error and display that errors message in the template??

File: packages/ember-glimmer/tests/integration/application/engine-test.js [hunk 2]
```diff
@@ -124,6 +139,220 @@ moduleFor('Application test: engine rendering', class extends ApplicationTest {
...
+  }
+
+  ['@test error route works for the application route of an Engine'](assert) {
+    this.setupAppAndRoutableEngine();
+    this.application.__registry__.resolver.moduleBasedResolver = true;
+    this.additionalEngineRegistrations(function() {
+      this.register('template:error', compile('Error!'));
+      this.register('route:post', Route.extend({
+        model() {
+          return RSVP.reject();
```
Comment (rwjblue) [inline packages/ember-glimmer/tests/integration/application/engine-test.js]: Same as above (reject with an error + show error message in template)

File: packages/ember-glimmer/tests/integration/application/engine-test.js [hunk 3]
```diff
@@ -124,6 +139,220 @@ moduleFor('Application test: engine rendering', class extends ApplicationTest {
...
+  }
+
+  ['@test error substate route works for a child route of an Engine'](assert) {
+    this.setupAppAndRoutableEngine();
+    this.application.__registry__.resolver.moduleBasedResolver = true;
+    this.additionalEngineRegistrations(function() {
+      this.register('template:post_error', compile('Error!'));
+      this.register('route:post', Route.extend({
+        model() {
+          return RSVP.reject();
```
Comment (rwjblue) [inline packages/ember-glimmer/tests/integration/application/engine-test.js]: Same as above (reject with an error + show error message in template)

File: packages/ember-glimmer/tests/integration/application/engine-test.js [hunk 4]
```diff
@@ -124,6 +139,220 @@ moduleFor('Application test: engine rendering', class extends ApplicationTest {
...
+  }
+
+  ['@test error route works for a child route of an Engine'](assert) {
+    this.setupAppAndRoutableEngine();
+    this.application.__registry__.resolver.moduleBasedResolver = true;
+    this.additionalEngineRegistrations(function() {
+      this.register('template:post.error', compile('Error!'));
+      this.register('route:post.comments', Route.extend({
+        model() {
+          return RSVP.reject();
```
Comment (rwjblue) [inline packages/ember-glimmer/tests/integration/application/engine-test.js]: Same as above (reject with an error + show error message in template)

**Review by rwjblue — CHANGES_REQUESTED** [review]
Can you update the various rejects to all reject with a value (preferably with an Error), and add expect counts to the new tests?

I'll try do another pass of review tomorrow or Monday, but this is looking very good. Thank you!

**Comment by trentmwillis** [comment]
@rwjblue addressed your comments and moved the tests to a more logical spot in that file.

**Comment by trentmwillis** [comment]
Also added some tests in ember-engines that pass after introducing the changes in this PR: https://github.com/dgeb/ember-engines/pull/247

**Review by rwjblue — APPROVED** [review]

**Comment by rwjblue** [comment]
Awesome, thank you!

**Comment by trentmwillis** [comment]
@rwjblue any chance this could get cherry picked into beta?

**Comment by rwjblue** [comment]
Yep, done!  (was on mobile when I merged originally)

**Comment by trentmwillis** [comment]
Thanks!

**Pull request merged**